search-app/src/cmr/search/api/concepts_search.clj

```diff
@@ -70,6 +80,74 @@
+  [headers]
+  (let [err-msg (str "The CMR does not currently allow querying across granules in all "
+                     "collections. To help optimize your search, you should limit your "
+                     "query using conditions that identify one or more collections, "
+                     "such as provider, provider_id, concept_id, collection_concept_id, "
+                     "short_name, version or entry_title. Visit the CMR Client Developer "
+                     "Forum at https://wiki.earthdata.nasa.gov/display/CMR/"
+                     "Granule+Queries+Now+Require+Collection+Identifiers for more "
+                     "information, and for any questions please contact "
+                     "support@earthdata.nasa.gov.")]  
+    (when (reject-all-granule-query? headers)
+      (svc-errors/throw-service-error :bad-request err-msg))))
+
+(defn- illegal-concept-id-in-granule-query?
+  "Check to see if any concept ids are not starting with G and C."
+  [concept-id-param]
+  (when (some? concept-id-param)
+    (let [concept-id-param-list (if (sequential? concept-id-param)
+                                  (remove #(= "" %) concept-id-param)
+                                  [concept-id-param])]
+      (some #(not (re-find #"^[GC]" %)) concept-id-param-list))))
+
+(defn empty-string-values? 
+  "This function is used by remove-map-keys function.
+   The keys will be removed if their values are empty strings, or if it's sequential and
+   only contain empty strings.
+   Note:  When you pass a parameter without = sign, it is considered a nil value and get  
+   filtered out before getting to the params; with the = sign, no value, it becomes empty 
+   string."
+  [v]
+  (or (= "" v)
+      (when (sequential? v)
+        (= [] (remove #(= "" %) v)))))
+
+(defn- all-granule-params?
+  "Returns true if it's a all granule query params.
+   Note: parameters with scroll-id don't need to be checked because it inherits the search
+   parameters from the original query, which would have been handled already." 
+  [scroll-id coll-constraints]
+  (and (not (some? scroll-id))
+       (empty? coll-constraints)))
+
+(defn- handle-granule-search-params
+  "Check the params when it is a granule search query."
+  [headers concept-type params scroll-id]
+  (when (= :granule concept-type)
+    ;; Check to see if any concept-id(s) are not starting with C or G. If so, bad request.
+    ;; otherwise, check to see if it's all-granule-params?, if so, handle it.
+    (let [params (lp/replace-parameter-aliases (util/map-keys->kebab-case params))
+          params (util/remove-map-keys empty-string-values? params)
+          constraints (select-keys params all-gran-validation/granule-limiting-search-fields)
+          concept-id-param (:concept-id constraints)
+          illegal-concept-id-msg (str "Granule query concept_id param [" concept-id-param
+                                      "] contains concept ids not starting with G or C.")]
```

Comment:
Invalid concept_id [concept-id-param]. For granule queries concept_id must be either a granule or collection concept ID.

Reply:
Done.